client/app/components/QuerySelector.jsx

```diff
@@ -0,0 +1,103 @@
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import cx from 'classnames';
+import Input from 'antd/lib/input';
+import { Query } from '@/services/query';
+import { QueryTagsControl } from '@/components/tags-control/TagsControl';
+
+const SEARCH_DEBOUNCE_DURATION = 200;
+
+function fetchRecent() {
+  return Query.recent().$promise
+    .then((results) => {
+      const filteredResults = results.filter(item => !item.is_draft); // TODO: filter at the fetch level
+      return Promise.resolve(filteredResults);
+    });
+}
+
+function search(q) {
+  return new Promise((resolve) => {
+    Query.query({ q }, ({ results }) => {
+      resolve(results);
+    });
+  });
+}
+
+export default function QuerySelector(props) {
+  const [searchTerm, setSearchTerm] = useState();
+  const [searching, setSearching] = useState();
+  const [searchResults, setSearchResults] = useState([]);
+  const [selectedQuery, setSelectedQuery] = useState();
+
+  // on search term changed, debounced
+  useEffect(() => {
+    setSearching(true);
+    const debounce = setTimeout(() => {
+      const promise = searchTerm ? () => search(searchTerm) : fetchRecent;
+      promise()
+        .then(setSearchResults)
+        .finally(() => {
+          setSearching(false);
+        });
+    }, SEARCH_DEBOUNCE_DURATION);
+
+    return () => {
+      clearTimeout(debounce);
+    };
+  }, [searchTerm]);
+
+  // on query selected
+  useEffect(() => {
+    const id = selectedQuery ? selectedQuery.id : null;
+    props.onChange(id);
+  }, [selectedQuery]);
```

Comment:
Feels like this should be in the `onClick` handler.

Another way to look at it: would you use `componentWillUpdate` for similar purpose?

Reply:
@arikfr `componentWillUpdate` is deprecated, but `useEffect` is a replacement for `componentDidUpdate` - so if you doubt you'll use cDU for this - keep in mind that `useEffect` _is_ cDU

Reply:
Since `<Select>` doesn't allow an object as an option value, the query.id is the value which is sent to `selectQuery`

```jsx
<select onChange={selectQuery}>
  <option value={query.id}>{query.name}</option>
</select>
```
which uses the id to find the query object and updates the state.
```js
function selectQuery(queryId) {
  const query = find(searchResults, { id: queryId });
  if (!query) { // shouldn't happen
    toastr.error('Something went wrong.. Couldn\'t select query');
  }
  setSelectedQuery(query);
}
```

An effect then reacts to the state change:
```js
useEffect(() => {
  setSearchTerm(selectedQuery ? null : ''); // empty string forces recent fetch
  props.onChange(selectedQuery);
}, [selectedQuery]);
```

Reply:
(The above is code from #3501)

Reply:
I probably miss something, but why not call `setSearchTerm` and `props.onChange` in `selectQuery`?

Reply:
Cause I wanted clearing it to use only `setSelectedQuery`.
https://github.com/getredash/redash/blob/0d209cb6dbd44195ceaf33bac56c684e061f6def/client/app/components/QuerySelector.jsx#L36

Reply:
Putting a condition in `selectQuery` feels like a cleaner approach, because what this handler in fact does is to select a query. Now you have this logic split into places: one in the handlers (selectQuery and this anonymous one) and some watcher.

Reply:
Done.
https://github.com/getredash/redash/blob/cd9f1e0d92fe92d4c6d09176e3fd497f44bff4c6/client/app/components/QuerySelector.jsx#L89-L101